core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java

```diff
@@ -0,0 +1,247 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.xcontent.ObjectParser.ValueType;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Adapts {@link ObjectParser} to work with object that have constructors who's arguments are mixed in with its other settings. Queries are
+ * like this, for example <code>ids</code> requires <code>types</code> but always parses the <code>values</code> field on the same level. If
+ * this doesn't sounds like what you want to parse have a look at
+ * {@link ObjectParser#declareNamedObjects(BiConsumer, ObjectParser.NamedObjectParser, Consumer, ParseField) which solves a slightly
+ * different but similar sounding problem.
+ * <p>
+ * Here is how you declare an ObjectSpooler:
+ * <p>
+ * <pre><code>
+ *   private static final ObjectSpooler<Thing, Void> PARSER = new ObjectSpooler<>("thing",
+ *           a -> new Thing((String) a[0], (String) a[1]));
+ *   static {
+ *       PARSER.declareString(PARSER.constructorArg(), new ParseField("animal"));
+ *       PARSER.declareString(PARSER.constructorArg(), new ParseField("vegetable"));
+ *       PARSER.declareInt(Thing::setMineral, new ParseField("mineral"));
+ *       PARSER.declareInt(Thing::setFruit, new ParseField("fruit"));
+ *   }
+ * </code></pre>
+ * <p>
+ * This does add some overhead compared to just using {@linkplain ObjectParser} directly. On a 2.2 GHz Intel Core i7 MacBook Air running on
+ * battery power in a reasonably unscientific microbenchmark it is about 100 microseconds for a reasonably large object, less if the
+ * constructor arguments are first. On this platform with the same microbenchmarks just creating the XContentParser is around 900
+ * microseconds and using {#linkplain ObjectParser} directly adds another 300 or so microseconds. In the best case
+ * {@linkplain ObjectSpooler} allocates two additional objects per parse compared to {#linkplain ObjectParser}. In the worst case it
+ * allocates <code>2 + 2 + 2 * param_count</code> objects per parse. If this overhead is too much for you then feel free to have
```

Comment:
I assume the figure here for the worst case is `2 + 2 + (2 * param_count)` rather than `(2 + 2 + 2) * param_count` but could you add the (unnecessary) brackets here to make it explicit and obvious which one it is